---
title: "MBKGKKMW_phase3"
author: "Magdalena Bogdańska, Karolina Gajewska, Katarzyna Kańska, Marcin Wojno"
date: "Linear and Mixed Models with Examples for Biological and Medical Data"
output: 
  html_document:
    toc: TRUE
---

# 1. Introduction
It is believed that in order to improve cancer care the key task is to understand genomics. The Cancer Genome Atlas (TCGA) is an international effort to collect and merge genetic and clinical data related to various types of cancer. All data generated by TCGA Research Network are made open to the public through the Data Coordinating Center and the TCGA Data Portal. The TCGA Data Portal [1] provides a platform for researchers to search, download, and analyze data sets generated by TCGA. It contains clinical information, genomic characterisation data and high level sequence analysis of the tumour genomes.

In this project we try to respond the question:

How is the expression of genes related to the type of cancer? 

We use statistical framework to find relevant genes and indicate differences of their expression.

# 2. Materials and Methods 
We base our analysis of gene expression on two datasets available in the RTCGA.PANCAN12 package for R [2]: one containing information on genes expression and the second one with clinical data.
The first dataset contains information of expression of 16115 genes for 3599 patients, while the second one provides data about characteristics of 5158 patient (type of tumour, age at diagnosis and gender). 

We aim to identify genes which expressed differently among cancer types using statistical test. The null hypothesis states that means of variables (genes expression) is the same for all groups (caner types). We reject this hypothesis if p-value is less than 0.05. In that case we claim that such gene is expressed differently among types of tumours (that is mean value in some groups is different from others).

Typically to such purpose we would use Anova test but as some assumptions of ANOVA are violated, we use non-parametric test (Kruskal-Wallis test). 

We analyse only those genes which p-values from Kruskal-Wallis test after Bonferroni adjustment are smaller than 0.05. 
First we select gene with smallest p-value. Second gene is one with lowest p-value from Kruskal-Wallis test selected out of genes which are weakly correlated with first gene. We continue such procedure and select weakly correlated genes which has the lowest p-values in Kruskal-Wallis test. 

We perform post-hoc tests (LSD tests) to verify which cancer types are different for genes selected by our procedure.

# 3. Results: differences in genes expression profiles for different cancer types
Names of selected genes are the following ones:
`?|10357`, `A1CF`, `A1BG`, `?|90288`, `?|8225`, `?|653553`, `?|645851`, `?|390284`.

Based on the outcomes of LSD tests we can observe that using only 3 genes, namely `?|10357`, `?|8225`, `?|390284`, one could differentiate all cancer types in this study. 
Gene `?|10357` differentiate endometrioid cancer, head and neck cancer, lung squamous cell carcinoma, bladder cancer, kidney clear cell carcinoma and ovarian cancer.
Gene `?|8225` differentiate colon cancer, rectal cancer, breast cancer, kidney clear cell carcinoma and ovarian cancer.
Gene `?|390284` differentiate endometrioid cancer, lung adenocarcinoma, acute myeloid leukemia and glioblastoma.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(ggplot2)
library(agricolae)

load("expression.rda")
load("clinical.rda")

ShowPlot <- function(gene, myPalette){
  df <- na.omit(data.frame(GeneExpression = as.numeric(t(expression[gene,])), CancerType = as.factor(t(clinical[1,]))))
  ggplot(df, aes(x = CancerType, y = GeneExpression, color = CancerType)) + geom_boxplot() +
    coord_flip() +
    scale_colour_manual(values=myPalette) + theme(legend.position="none") +
    ggtitle(paste0("Gene ", gene))
}

myPalette <- c("#E69F00", "#000000", "#000000", "#000000", "#56B4E9", "#009E73", "#000000", "#000000", "#8B13D1", "#000000", "#0072B2", "#000000")
ShowPlot("?|390284", myPalette)
```

Colouring is based on LSD test results.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
df <- na.omit(data.frame(GeneExpression = as.numeric(t(expression["?|390284",])), CancerType = as.factor(t(clinical[1,]))))
(LSD.test(aov(GeneExpression ~ CancerType, data=df),"Cancer"))$groups
```

In such way we can indicate small number of genes characteristic for each type of cancer in this study. 
We summarise results in a following table:

Cancer type                   |     Genes
------------------------------+----------------------------
Acute Myeloid Leukemia        |     `?|645851`, `?|390284`
Bladder Cancer                |     `?|10357`, `?|653553`
Breast Cancer                 |     `A1BG`, `?|8225`, `?|653553` 
Colon Cancer                  |     `?|8225`
Endometrioid Cancer           |     `?|10357`, `?|90288`, `?|390284`
Glioblastoma                  |     `A1CF`, `?|645851`, `?|390284`
Head and Neck Cancer          |     `?|10357`, `A1BG`, `?|653553`, `?|645851` 
Kidney Clear Cell Carcinoma   |     `?|10357`, `A1BG`, `?|653553`
Lung Adenocarcinoma           |     `?|653553`, `?|390284`
Lung Squamous Cell Carcinoma  |     `?|10357`
Ovarian Cancer                |     `?|10357`, `?|8225`, `?|653553`, `?|390284`
Rectal Cancer                 |     `?|8225`

# 4. Conclusions

We indicate very small number of genes which can identify types of cancers considered in this study. Apart of that for each cancer type we have provided a small list of genes which seems to have characteristic values to determine cancer type.

# 5. Literature

[1] cancergenome.nih.gov

[2] github.com/mi2-warsaw/RTCGA.data/tree/master/RTCGA.PANCAN12

[3] Analiza danych z programem R. Modele liniowe z efektami stałymi, losowymi i mieszanymi, P. Biecek, Wydawnictwo Naukowe PWN 2011

[4] The Cancer Genome Atlas (TCGA): an immeasurable source of knowledge, K. Tomczak, P. Czerwi?ska and M. Wiznerowicz, Contemp Oncol (Pozn). 2015; 19(1A): A68?A77, doi: 10.5114/wo.2014.47136

# A. Appendix 
Here we provide code in R we used to extract results described above.

## A.1. Data preprocessing
In both datasets we get rid of data on samples in the smallest group "TCGA Formalin Fixed Paraffin-Embedded Pilot Phase II" and save it.
```{r, warning=FALSE, message=FALSE, eval=FALSE}
load("expression.rda")
load("clinical.rda")
expression <- expression[,clinical[1,]!="TCGA Formalin Fixed Paraffin-Embedded Pilot Phase II"]
clinical <- clinical[,clinical[1,]!="TCGA Formalin Fixed Paraffin-Embedded Pilot Phase II"]
save(expression, file = "expression.rda")
save(clinical, file="clinical.rda")
```
## A.2. Test for relations between gene expressions with cancer types
In previous steps of project using Shapiro test for normality and Levene test for homogeneity of variance among groups we verified that expression of many genes does not follow assumptions of Anova. Thus in order to treat each variable in the same way, we perform Kruskal-Wallis test for all genes.
After Bonferroni adjustment we stay with 16104 genes which seems to be significantly different across cancer types. However from the point of view of possible applications we are interested in samll number of such significant genes.

```{r, warning=FALSE, message=FALSE, eval=FALSE}
PerformKruskal <- function(gene){
  cancerData <- na.omit(data.frame ( gene, as.factor(t(clinical[1,]))))
  colnames(cancerData) <- c("gene", "cancer")
  tryCatch({
    kruskal.test(gene ~ cancer, data = cancerData)$p.value
  }, error=function(e){c(NA, NA)})
}
p.values.k <- t(apply(expression, 1, PerformKruskal))
save(p.values.k, file = "p.values.k.rda")
Pvals<-as.data.frame(t(p.values.k))
summary(Pvals$V1 > 0.05)
```
We can observe that using standard p-value equal to 0.05 all genes are susceptible to express in a different way for different cancer types. 
In order to account for problems with multiple hypotheses testing, we perform adjustments of p-values. 
We compute adjustment of p-values with different methods built in R (Bonferroni, BH, Holm, Hochberg, Hommel, BY). Only using Bonferroni method the influence of some genes (11) wasn't statistically significant. Here we present procedure for Bonferroni method.
```{r, warning=FALSE, message=FALSE, eval=FALSE}
Pvals$Bonferroni = p.adjust(Pvals$V1, method = "bonferroni")
summary(Pvals$Bonferroni > 0.05)
save(Pvals, file = "p.values.k.p.Bonferroni.rda")
```
## A.3. Procedure of selecting the most significant genes 

We take into consideration only those genes which p-values from Kruskal-Wallis test after Bonferroni adjustment are smaller than 0.05. 
First we select gene with smallest p-value. In the following analysis we reject genes which expression profiles are correlected with the expression of this first gene.
Second gene is one with lowest p-value from Kruskal-Wallis test selected out of genes which are weakly correlated with first gene. We continue such procedure and select weakly correlated genes which has the lowest p-values in Kruskal-Wallis test. Due to the computational costs we choose 8 genes and later analyse their expression.

```{r, warning=FALSE, message=FALSE, eval=FALSE}
s<-8
genes1<-rownames(Pvals)[which(Pvals[,2]<0.05)] #16104 genes with small p-values after Bonferroni adjustment
genes<-matrix(NA,nrow=length(genes1),ncol=s)
genes[,1]<-genes1

best_genes<-rep(NA,s)
best_genes[1]<-rownames(Pvals)[which.min(Pvals[genes[,1],2])] #?|10357  
save(best_genes, file = "best_genes.rda")

ComputeCorr<-function(gene,best) {
  cor(as.numeric(expression[gene,]), as.numeric(expression[best,]), use="pairwise.complete.obs")
}

for(i in 2:s){
  genes_temp<-genes[which(!is.na(genes[,i-1])),i-1] 
  
  Corr<-sapply(genes_temp,function(x) ComputeCorr(x,best_genes[i-1]))
  Corr<-as.data.frame(Corr)
  rownames(Corr)<-genes_temp
  save(Corr, file = paste0("corr.",i-1,".rda"))
  
  genes2<-rownames(Corr)[which(abs(Corr)<0.5)] 
  genes[,i]<- c(genes2, rep(NA, times=length(genes[,1])-length(genes2)))

  best_genes[i]<-rownames(Pvals)[which.min(Pvals[genes[,i],2])]
  save(best_genes, file = "best_genes.rda")
}
```
In order to confirm the significance of genes selected we perform post-hoc tests. The proper post-hoc test in this case is LSD test or Scheffe test as they don't require that groups have equal sizes. Here we perform LSD tests.

We can visualize differences in genes expressions using ggplots. An examplary code for one gene is provided below.
```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(ggplot2)
library(agricolae)
best_genes<-c("?|10357", "A1CF", "A1BG", "?|90288", "?|8225", "?|653553", "?|645851", "?|390284")

DataSelected<- data.frame(as.factor(t(clinical[1,])),t(expression[c(best_genes),]))
colnames(DataSelected)[1]<-"Cancer"
colnames(DataSelected)
levels(DataSelected$Cancer)<-c("Acute Myeloid Leukemia","Bladder Cancer","Breast Cancer","Colon Cancer","Endometrioid Cancer","Glioblastoma","Head and Neck Cancer","Kidney Clear Cell Carcinoma","Lung Adenocarcinoma","Lung Squamous Cell Carcinoma","Ovarian Cancer","Rectal Cancer")
  
ggplot(DataSelected, aes(Cancer,X..10357,fill = Cancer)) + geom_boxplot() + coord_flip()
(LSD.test(aov(X..10357 ~ Cancer),"Cancer"))$groups
```  

## A.4 Plots for all selected genes

Here we present a comparison of histogram for all selected genes. The type of cancer for which where LSD test indicated siginicant difference in mean are coloured. 

```{r, warning=FALSE, message=FALSE}
myPalette <- c("#000000", "#E69F00", "#000000", "#000000", "#56B4E9", "#000000", "#009E73", "#8B13D1", "#000000", "#0072B2", "#D55E00", "#000000")
ShowPlot("?|10357", myPalette)

myPalette <- c("#000000", "#000000", "#000000", "#000000", "#CC2B84", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000")
ShowPlot("A1CF", myPalette)

myPalette <- c("#000000", "#000000", "#E69F00", "#000000", "#000000", "#000000", "#56B4E9", "#009E73", "#000000", "#000000", "#000000", "#000000")
ShowPlot("A1BG", myPalette)

myPalette <- c("#000000", "#000000", "#000000", "#000000", "#000000", "#CC2B84", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000")
ShowPlot("?|90288", myPalette)

myPalette <- c("#000000", "#000000", "#E69F00", "#56B4E9", "#009E73", "#000000", "#000000", "#000000", "#000000", "#000000", "#8B13D1", "#0072B2")
ShowPlot("?|8225", myPalette)

myPalette <- c("#000000", "#E69F00", "#56B4E9", "#000000", "#000000", "#000000", "#009E73", "#8B13D1", "#0072B2", "#000000", "#D55E00", "#000000")
ShowPlot("?|653553", myPalette)

myPalette <- c("#E69F00", "#000000", "#000000", "#000000", "#000000", "#56B4E9", "#009E73", "#000000", "#000000", "#000000", "#000000", "#000000")
ShowPlot("?|645851", myPalette)

myPalette <- c("#E69F00", "#000000", "#000000", "#000000", "#56B4E9", "#009E73", "#000000", "#000000", "#8B13D1", "#000000", "#0072B2", "#000000")
ShowPlot("?|390284", myPalette)
```
