---
  title: "Project 1"
author: "Magdalena Bogda?ska, Karolina Gajewska, Katarzyna Ka?ska, Marcin Wojno"
date: "Linear and mixed models with examples for biological and medical data"
output: 
  html_document:
  toc: TRUE
---
  
  ## 1. Introduction
  
  It is believed that in order to improve cancer care the key task is to understand genomics [2]. 
The Cancer Genome Atlas (TCGA) is an international effort to collect and merge genetic and clinical data related to various types of cancer. 
All data generated by TCGA Research Network are made open to the public through the Data Coordinating Center and the TCGA Data Portal.
The TCGA Data Portal [1] provides a platform for researchers to search, download, and analyze data sets generated by TCGA. It contains clinical information, genomic characterisation data and high level sequence analysis of the tumour genomes. 

Having such comprehensive repository one can look for similarities and differences across cancers, not only in each cancer separately. 
One can notice that the expression of some genes is different in groups of patients with different kinds of tumours.  
In this project we will try to respond the question: 
  
  Is expression of genes in different patients related to the kind of patient tumour? 

We will use statistical framework to find relevant genes and measure differences of their expression.

-------------------------------------------------------------------------------------------
  ## 2. Data preprocessing
  We will base our analysis of gene expression on two datasets available in the RTCGA.PANCAN12 package for R loaded in GitHub [3]: one containing information on genes expression and the second one with clinical data.
The first dataset is divided and stored in two matrices due to GitHub limits. Thus first we load and bind both files with gene expression profiles data.
```{r, warning=FALSE, message=FALSE}
load("expression.cb1.rda")
load("expression.cb2.rda")
expression <- rbind(expression.cb1, expression.cb2)
dim(expression) 
expression[1:4,1:4]
```
Rows correspond to selected 16115 genes while columns correspond to 3599 patients (as the first column contains names of genes). Name of the column is an identifier of the patient.

Now we load clinical data.
```{r, warning=FALSE, message=FALSE}
load("clinical.cb.rda")
clinical<-clinical.cb[,c("sampleID", "X_cohort", "age_at_initial_pathologic_diagnosis", "gender")]
dim(clinical)
clinical[1:5,]
```
5158 rows correspond to each patient. We select 4 columns with an ID of the patients, type of tumour, age at diagnosis and gender. 
TCGA has chosen cancers for study based on specific criteria that include:
  1) poor prognosis and overall public health impact,
2) availability of human tumor and matched-normal tissue samples.
In RTCGA.PANCAN12 package there is available data on the following types of cancer:
  ```{r, warning=FALSE, message=FALSE}
levels(clinical$X_cohort)
```

As we deal with big matrices, we try to simplify them as much as possible. We will name the row headings with patients ID in clinical data and with names of genes in data frame "expression". Subsequently we get rid of first columns in both data frames.

```{r, warning=FALSE, message=FALSE}
rownames(clinical) <- clinical[,1] 
clinical <- clinical[,2:ncol(clinical)] 
dim(clinical) 

rownames(expression) <- expression[,1] 
expression <- expression[,2:ncol(expression)]
dim(expression)
```

We also observe that names of pateints indata frame "expression" have dashes instead of dots (as in data frame "clinical"). We wil change it with command "str_replace_all" from package "stringr". We select from data frame "clinical" only rows for those patients which genes expression prolifes were described in data frame "expression". We export both datasets.

```{r, warning=FALSE, message=FALSE}
library(stringr)
colnames(expression)<-str_replace_all(colnames(expression),"[.]","-") 
expression[1:3,1:3]

#clinical <- clinical[colnames(expression),] 
#clinical[1:5,1:3]

temp <- clinical[colnames(expression),]
clinical <- as.data.frame(t(temp))

# save(expression, file = "expression.rda")
# save(clinical, file = "clinical.rda")
```

## 3. Analysis of differences in gene expression for different types of tumours
We use analysis of variance (ANOVA) for each gene to identify genes which expressed differently in total cohort of patients. 
We reject the zero hypothesis if p-value is greater than 0.5. In that case we will claim that such gene is expressed differently among groups of tumours (that is mean value in some group is different from others).

We can observe that cohorts of each cancer type are not equal in size. We predict that some assumptiont of ANOVA may be violated, to be specific we will verify normality using Shapiro-Wilk test. P-values for ANOVA test and Shapiro-Wilk test of standardized residuals will be saved in vector "p.values". If data on gene don't pass normality test, we will preform Box-Cos transformation.

We will present testing procedure for first gene.
```{r, warning=FALSE, message=FALSE}
summary(as.factor(t(clinical)[,1]))

library(MASS)

p.values <- as.data.frame(matrix(-1, nrow=nrow(expression), ncol=2))
colnames(p.values) <- c("anova", "normality")

load("expression.rda")
load("clinical.rda")

i=1
cancerData <- na.omit(data.frame ( t(expression[i,]), as.factor(t(clinical[1,]))))
colnames(cancerData) <- c("gene", "cancer")

library("ggplot2")
ggplot(cancerData, aes(x=gene)) + geom_histogram(binwidth=.1, colour="black", fill="white")
ggplot(aes(y = gene, x = cancer, fill = cancer), data = cancerData) + geom_boxplot()

if (shapiro.test(t(expression[1,]))$p.value >= 0.05)
{
model<-anova(lm(gene ~ cancer, data = cancerData))
p.values$anova[i]<-model[1,5]
p.values$normality[i] <- shapiro.test(rstandard(model))$p.value
} else {
if (min(cancerData$gene)<0) 
cancerData$gene <- cancerData$gene - min(cancerData$gene) + 1

tmp <- boxcox(gene ~ cancer, data = cancerData, lambda=seq(-2, 2, by=.01), plotit = FALSE)
lambda <- tmp$x[which.max(tmp$y)]

if (lambda == 0){
cancerData$gene <- log(cancerData$gene)
} else {
cancerData$gene <- (cancerData$gene^lambda - 1)/lambda
}

model<-anova(lm(gene ~ cancer, data = cancerData))
p.values$anova[i]<-model[1,5] 
p.values$normality[i] <- shapiro.test(rstandard(model))$p.value
}

p.values$anova[i] <= 0.05
p.values$normality[i] > 0.05

ggplot(cancerData, aes(x=gene)) + geom_histogram(binwidth=.1, colour="black", fill="white")
ggplot(aes(y = gene, x = cancer, fill = cancer), data = cancerData) + geom_boxplot()
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
for (i in 1:nrow(expression)){
print(i)

cancerData <- na.omit(data.frame ( t(expression[i,]), as.factor(t(clinical[1,]))))
colnames(cancerData) <- c("gene", "cancer")

tryCatch({

if (shapiro.test(t(expression[1,]))$p.value >= 0.05){
model<-anova(lm(gene ~ cancer, data = cancerData))
p.values$anova[i]<-model[1,5]
p.values$normality[i] <- shapiro.test(rstandard(model))$p.value
} else {
if (min(cancerData$gene)<0) 
cancerData$gene <- cancerData$gene - min(cancerData$gene) + 1

tmp <- boxcox(gene ~ cancer, data = cancerData, lambda=seq(-2, 2, by=.01), plotit = FALSE)
lambda <- tmp$x[which.max(tmp$y)]

if (lambda == 0) {
cancerData$gene <- log(cancerData$gene)
} else {
cancerData$gene <- (cancerData$gene^lambda - 1)/lambda
}

model<-anova(lm(gene ~ cancer, data = cancerData))
p.values$anova[i]<-model[1,5]

p.values$normality[i] <- shapiro.test(rstandard(model))$p.value
}
}, error=function(e){print("failed")})

}
```

```{r, warning=FALSE, message=FALSE}
load("pvalues.rda")
sum(as.numeric(p.values$anova <= 0.05))
sum(as.numeric(p.values$normality > 0.05))
```

## 4. Results
After performing ANOVA for all genes qwe obtain that means for all 16115 genes are different among groups of cancers while all residuals are not normally distributed with significance level 5%).

It is natural to pose a question then: 

Is it intuitial that so many genes are different for tumours? Is it something that we would predict? 

We should face that cancer is a disease of the genome and as more is learned about cancer tumours, the more we are finding that each tumour has its own set of genetic changes [1]. 

We shall remember the enormous level of complexity of tumours. From 2011 tumours have been defined based on six hallmarks which describe 
biological capabilities acquired during the multistep development of tumours. They include:

- sustaining proliferative signaling, 

- evading growth suppressors, 

- resisting cell death, 

- enabling replicative immortality, 

- inducing angiogenesis, 

- activating invasion and metastasis [5]. 

Apart of that in human tumours often occurs the reprogramming of energy metabolism and evading immune destruction. In addition to cancer cells, tumors exhibit another dimension of complexity: they recruit normal cells that contribute to the tumour growth by creating the "tumour microenvironment." Underlying all these hallmarks there is a genome instability.


## 5. Literature
[1] cancergenome.nih.gov

[2] RTCGAToolbox: A New Tool for Exporting TCGA Firehose Data, M. K. Samur, Plos One 2014; doi: 10.1371/journal.pone.0106397

[3] github.com/mi2-warsaw/RTCGA.data/tree/master/RTCGA.PANCAN12

[4] The Cancer Genome Atlas (TCGA): an immeasurable source of knowledge, K. Tomczak, P. Czerwi?ska and M. Wiznerowicz, Contemp Oncol (Pozn). 2015; 19(1A): A68?A77, doi: 10.5114/wo.2014.47136

[5] Hallmarks of cancer: the next generation, D. Hanahan and R. A. Weinberg, Cell. 2011; 144(5):646-74. doi: 10.1016/j.cell.2011.02.013.

